### PR TITLE
Channel attack

### PR DESCRIPTION
**Pull request by cyberj0g**
Addresses #115 
Changes:
- Verification is performed on frames with all channels preserved
- Chi-squared distance is estimated over channel value histograms to defend against channel manipulation in conjunction with other metrics
- New metric is calculated on full renditions dataset
- Models are trained with the new metric

### CODE REVIEW

File: feature_engineering/update_dataset.py [hunk 1]
```diff
@@ -3,19 +3,22 @@
...
+import re
 import sys
 import os
 import shutil
 import time
 import numpy as np
 import pandas as pd
 import tqdm
+import json
+from machine_learning.cnn.image_dataset_generator import PairWriter
```
Comment (yondonfu) [inline feature_engineering/update_dataset.py:15]: What does the PairWriter do?  I can't seem to find the PairWriter code that is being imported here.

File: scripts/asset_processor/video_asset_processor.py [hunk 2]
```diff
@@ -200,19 +205,13 @@ def capture_to_array(self, capture):
             width = frame_data.frame.shape[0]
             pixels += height * width
 
+            if not self.markup_master_frames and self.image_pair_callback is not None:
+                self.image_pair_callback(self.master_samples_hd[i], frame_data.frame, len(frame_list), ts_diff, self.original_path, capture.filename)
```
Comment (yondonfu) [inline scripts/asset_processor/video_asset_processor.py:209]: What is the purpose of the image pair callback?

File: scripts/modeling_toolbox/utils.py [hunk 3]
```diff
@@ -13,14 +13,10 @@ def load_data(data_uri, nrows=None):
...
     data_df.rename(lowercase, axis='columns', inplace=True)
-    data_df.drop(['dimension_x'], axis=1, inplace=True)
     data_df.rename(columns={'attack':'rendition',
-                            'title':'source',
-                            'dimension': 'dimension_x'},
+                            'title':'source'},
                    inplace=True)
     data_df['rendition'] = data_df['rendition'].apply(lambda x: set_rendition_name(x))
-    data_df['dimension_y'] = data_df['rendition'].apply(lambda x: int(x.replace('p','').split('_')[0]))
-    data_df['size_dimension_ratio'] = data_df['size'] / (data_df['dimension_y'] * data_df['dimension_x'])
```
Comment (yondonfu) [inline scripts/modeling_toolbox/utils.py:23]: Why are we removing `size_dimension_ratio`?

File: scripts/asset_processor/video_metrics.py [hunk 4]
```diff
@@ -468,14 +467,14 @@ def compute_metrics(self,
 
         rendition_metrics = {}
         # Some metrics only need the luminance channel
-        reference_frame_gray = reference_frame
-        rendition_frame_gray = rendition_frame
-        next_reference_frame_gray = next_reference_frame
-        next_rendition_frame_gray = next_rendition_frame
+        reference_frame_v = cv2.cvtColor(reference_frame, cv2.COLOR_BGR2HSV)[..., -1]
```
Comment (yondonfu) [inline scripts/asset_processor/video_metrics.py:470]: Just to make sure that I understand correctly:

We're still computing most of the metrics using just the V component by converting the frame from BGR to HSV here, but we're computing certain metrics such as `temporal_histogram_distance` using all color channels from the BGR frame? And in order to accomplish this we pass the BGR frames to this VideoMetrics class so the VideoMetrics class can do the HSV conversion instead of doing the HSV conversion and then passing the HSV frames to the VideoMetrics class?

Comment (cyberj0g) [reply scripts/asset_processor/video_metrics.py:470]: Yes, that's right. Full image metric works in conjunction with single-channel metrics.

File: scripts/modeling_toolbox/utils.py (hunk 3 above)
Comment (cyberj0g) [reply scripts/modeling_toolbox/utils.py:23]: This code was for previous version of VideoAssetProcessor, which didn't calculate some meta-features.

File: scripts/asset_processor/video_asset_processor.py (hunk 2 above)
Comment (cyberj0g) [reply scripts/asset_processor/video_asset_processor.py:209]: Mainly for debug purposes, to visualize or save pairs of matched frames. May also be useful for running some custom verification logic.

File: testing/mobilenet_test.py [hunk 5]
```diff
@@ -0,0 +1,5 @@
+checkpoint_name = 'mobilenet_v2_1.0_224' #@param
+checkpoint = checkpoint_name + '.ckpt'
+
+import tensorflow as tf
```
[the remark this answers is not in the document]
Comment (cyberj0g) [reply testing/mobilenet_test.py:4]: Removed.

File: testing/tests/test_var_fps.py [hunk 6]
```diff
@@ -44,9 +44,10 @@ def test_classification(self):
         rendition_dirs = [
             ('../data/renditions/720p_watermark/', True),
             ('../data/renditions/720p_60-24fps/', False),
+            #('../data/renditions/720p/', False),
         ]
         files = None
-        debug = False
+        debug = True
```
[the remark this answers is not in the document]
Comment (cyberj0g) [reply testing/tests/test_var_fps.py]: No, switched it off.

File: testing/tests/test_var_fps.py [hunk 7]
```diff
@@ -44,9 +44,10 @@ def test_classification(self):
         rendition_dirs = [
             ('../data/renditions/720p_watermark/', True),
             ('../data/renditions/720p_60-24fps/', False),
+            #('../data/renditions/720p/', False),
```
[the remark this answers is not in the document]
Comment (cyberj0g) [reply testing/tests/test_var_fps.py]: Removed.

File: feature_engineering/update_dataset.py (hunk 1 above)
Comment (cyberj0g) [reply feature_engineering/update_dataset.py]: It writes frame pairs matched by VideoAssetProcessor to disk. Moved the class to file.

File: scripts/modeling_toolbox/utils.py (hunk 3 above)
Comment (yondonfu) [reply scripts/modeling_toolbox/utils.py:23]: Ah I see and VideoAssetProcessor now [computes size_dimension_ratio](https://github.com/livepeer/verification-classifier/blob/8f48463dd6b1cd9def1ec19378562781af6e9e8e/scripts/asset_processor/video_asset_processor.py#L451) so that should already be included in the data read from a file right?

**Comment by yondonfu** [comment]
I'm seeing a failing test after following the steps for running unit tests included in the README

```
====================================================================== short test summary info ======================================================================
FAILED testing/tests/test_var_fps.py::TestVarFps::test_classification - assert (3 / 5) >= 0.8
======================================================== 1 failed, 5 passed, 5 warnings in 93.70s (0:01:33) =========================================================
```

I downloaded and extracted the CSV dataset and YT8M renditions mini dataset into the `../data` directory.

File: verifier/verifier.py [hunk 8]
```diff
@@ -229,26 +223,29 @@ def retrieve_models(self, uri, model_dir):
         """
         Function to obtain pre-trained model for verification predictions
         """
-        with file_locker.FileLocker('model_op.lock'):
```
Comment (yondonfu) [inline verifier/verifier.py:232]: Don't we need this file lock when gunicorn uses multiple processes?

File: verifier/verifier.py [hunk 9]
```diff
@@ -106,16 +105,11 @@ def pre_verify(self, source, rendition):
...
-        UL classifier has a higher TPR but lower TNR, meaning it is less restrictive towards
-        tampered assets. SL classifier has higher TNR but is too punitive, which is undesirable,
-        plus it requires labeled data.
+        The goal is to reduce the number of False Positives (tamper) to prevent wrongfully penalizing transcoder nodes. OCSVM model is expected to have higher precision (low FP) on novel data.
+        If OCSVM predicts the observation is an inlier (not tampered), we'll go with it, otherwise we'll use supervised model output.
         """
-        meta_condition = row['ul_pred_tamper'] == 1 and row['sl_pred_tamper'] == 1
-        if meta_condition:
-            return row['ul_pred_tamper']
+        if row['ul_pred_tamper'] == 0:
```
Comment (yondonfu) [inline verifier/verifier.py:111]: Why the change in the meta model condition?

Comment (cyberj0g) [reply verifier/verifier.py:111]: I believe it was incorrect before, because of confusion of what's the return of OCSVM model in case it predicts tamper (it returns -1 for an outlier, and 1 for inlier). The prediction is now converted to match 1=Tamper notation. Regarding the specific condition, we trust OCSVM model, if it detects an outlier, and use supervised Catboost model in all other cases. It gives additional 0.01 precision on test data.
From the comment:
        The goal is to reduce the number of False Positives (tamper) to prevent wrongfully penalizing transcoder nodes. OCSVM model is expected to have higher precision (low FP) on novel data.
        If OCSVM predicts the observation is an inlier (not tampered), we'll go with it, otherwise we'll use supervised model output.

File: verifier/verifier.py (hunk 8 above)
Comment (cyberj0g) [reply verifier/verifier.py]: Fixed.
